server/src/test/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesActionTests.java

```diff
@@ -0,0 +1,103 @@
+
+    private final ThreadPool threadPool = new TestThreadPool(getClass().getName());
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        ThreadPool.terminate(threadPool, 10, TimeUnit.SECONDS);
+    }
+
+    public void testCCSCompatibilityCheck() throws Exception {
+        Settings settings = Settings.builder()
+            .put("node.name", TransportSearchAction.class.getSimpleName())
+            .put(SearchService.CCS_VERSION_CHECK_SETTING.getKey(), "true")
+            .build();
+        ActionFilters actionFilters = mock(ActionFilters.class);
+        when(actionFilters.filters()).thenReturn(new ActionFilter[0]);
+        try {
+            TransportService transportService = MockTransportService.createNewService(Settings.EMPTY, Version.CURRENT, threadPool);
+
+            FieldCapabilitiesRequest fieldCapsRequest = new FieldCapabilitiesRequest();
+            fieldCapsRequest.indexFilter(new DummyQueryBuilder() {
+                @Override
+                protected void doWriteTo(StreamOutput out) throws IOException {
+                    if (out.getVersion().before(Version.CURRENT)) {
+                        throw new IllegalArgumentException("This query isn't serializable to nodes before " + Version.CURRENT);
+                    }
+                }
+            });
+
+            IndicesService indicesService = mock(IndicesService.class);
+            when(indicesService.getAllMetadataFields()).thenReturn(Collections.singleton("_index"));
+            ClusterService clusterService = new ClusterService(
+                settings,
+                new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS),
+                threadPool
+            );
+            TransportFieldCapabilitiesAction action = new TransportFieldCapabilitiesAction(
+                transportService,
+                clusterService,
+                threadPool,
+                actionFilters,
+                indicesService,
+                null
+            );
+
+            IllegalArgumentException ex = expectThrows(
+                IllegalArgumentException.class,
+                () -> action.doExecute(null, fieldCapsRequest, new ActionListener<FieldCapabilitiesResponse>() {
+                    @Override
+                    public void onResponse(FieldCapabilitiesResponse response) {}
+
+                    @Override
+                    public void onFailure(Exception e) {}
+                })
+            );
+
+            assertEquals(
+                "parts of writeable [org.elasticsearch.action.fieldcaps.FieldCapabilitiesRequest/unset] "
+                    + "are not compatible with version 8.0.0 and the 'search.check_ccs_compatibility' setting is enabled.",
+                ex.getMessage()
+            );
+            assertEquals("This query isn't serializable to nodes before " + Version.CURRENT, ex.getCause().getMessage());
```

Comment:
thanks, this is readable. I only wonder which part of this will be seen in Kibana, meaning is it correct that the most readable is in the cause, or should it be the other way around?